### ISSUE TITLE
Change the `std::error::Error` condition to `std` or Rust version >= 1.81

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3317]
![image](https://github.com/user-attachments/assets/6f3150f3-12b4-463d-a35c-e87614632bf8)

We'll need a sensible macro for this since we don't use `cfg` for Rust version any longer and also the impls differ in the trait crate name.

I think something like this:

```rust
impl_std_error! {
    impl Error for Foo {}
}
```

Where we inject `core::error::Error` in the Rust version >= 1.81 branch and `std::error::Error` in the else branch.

We also need to modify `write_err` to take the Rust version into account.

**Comment by apoelstra** [comment]
Oh nice! I had been a little frustrated by this because I don't think there's any way we can *bound* things on `core::error::Error` (nor can we box `core::error::Error` in any useful way) with a cfg.....but we can totally conditionally implement it on our types.

Hooray!

**Comment by Kixunil** [comment]
Changing the bound from `std::error::Error` to `core::error::Error` in Rust >= 1.81 should be fine but maybe it'll be a bit fiddly within the macro.

> nor can we box `core::error::Error` in any useful way

We can with `alloc` but I don't think we want to in most cases.

**Comment by apoelstra** [comment]
> We can with `alloc` but I don't think we want to in most cases.

Hmm, maybe if we had a `Box<dyn Display + Debug + 'static>` in pre-1.81.0, and a `Box<dyn Error>` post 1.81.0, that would work. But my worry is that we'd be changing the API in a non-additive way for people with a recent compiler.

**Comment by Kixunil** [comment]
I don't think those boxes are fully compatible because they have different vtables, so you can't pass a function returning it to  a function expecting a function returning `Box<dyn Display + Debug + 'static>`. We'd have to return an opaque type that implements the traits and also has conversions to the boxes.

**Comment by apoelstra** [comment]
> But really, I hope we won't have any `Box<dyn ...>` errors in the crate.

Yeah, I don't think we'll need them. We haven't had any cases here where error types wound up having generics on them (which is so unergonomic it makes you want to reach for a box..) nor cases where error types have become unreasonably large (though we haven't made any systemic effort to measure them).

In rust-miniscript we have both issues, so we're gonna have boxed errors there.

**Comment by Kixunil** [comment]
I've realized this is a breaking change because it changes formatting of the errors which some applications might rely on. We need to get it in 1.0.

**Comment by apoelstra** [comment]
Formatting of errors is not part of our public API (for semver purposes -- I do agree that we should avoid gratuitously changing them).

I'm sure we've discussed this elsewhere but I'm saying it again now.

**Comment by Kixunil** [comment]
I'd say things like exact wording are not part of public API but changing "foo failed" to "foo failed: bar is invalid" is very disrupting because then all downstream crates must adapt accordingly (especially if they test formatting of errors). I actually do have this problem in `ln-types` because `secp256k1` doesn't support `core::error::Error` but I want to support it soon. (I plant to make PR and not bitch about it, just flagging that this is important.)

**Comment by apoelstra** [comment]
> changing "foo failed" to "foo failed: bar is invalid" is very disrupting

We have this issue now with our `write_err` macro, right, even without trying to make it compiler-version-dependent?

BTW, if the answer to all this is "just bump MSRV to 1.80" I can probably live with that.

**Comment by Kixunil** [comment]
That one is very reliably dependent on whether `std` is turned on or not, which is easy to take care of downstream.

> BTW, if the answer to all this is "just bump MSRV to 1.80" I can probably live with that.

It's much less disruptive, just add another `#[cfg]`. I even found a nice pattern to make it easy in files containing many `Error` impls.

**Comment by apoelstra** [comment]
> It's much less disruptive, just add another `#[cfg]`. I even found a nice pattern to make it easy in files containing many `Error` impls.

What do you mean by this? We'd have another cfg flag that users would need to set in their `RUSTFLAGS`?

**Comment by Kixunil** [comment]
Not at all, our build script already does version detection, so we use that one.

**Comment by apoelstra** [comment]
Now I'm confused. I thought this is what you were complaining about above when you said it'd be a "breaking change"?

**Comment by Kixunil** [comment]
It's breaking to release a version that falls back to displaying sources and then changing it to a version that always displays sources. Of course some `no_std` users could experience breakage after upgrading the compiler version but I think that will happen regardless of what we do.

**Comment by apoelstra** [comment]
Ok, I think I've misunderstood your point. I think we agree on what to do and that it's a breaking change (sorta -- I continue to think that we'd be technically ok to do this breakage in a semver-minor version). Your point is that we need to do it *now, before 1.0*.

I had misunderstood you to mean that we should *avoid doing this* because it was breaking. And had thought that you were trying to find alternate things to do.

**Comment by Kixunil** [comment]
Correct. And it probably wouldn't be a huge issue if we changed it later but I think it'd be quite a dick move not add one reasonably simple change. (Though annoyingly large-scale I guess.)
